### DISCUSSION TITLE
Use generic instead of Any in Depends function

### DISCUSSION THREAD

**Discussion by invokermain** [fastapi/fastapi#6377] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Depends, FastAPI

app = FastAPI()


def dependency() -> int:
    return 99


@app.get("/")
def read_root(my_int: int = Depends(dependency)):
    return {"an_int": my_int}
```

### Description

Is there any reason that Depends does not use generics to automatically type hint the injected dependency?

Currently, the  Depends function's return is typed as Any:

```python
def Depends(  # noqa: N802
    dependency: Optional[Callable[..., Any]] = None, *, use_cache: bool = True
) -> Any:
    return params.Depends(dependency=dependency, use_cache=use_cache)
```

This leads to a lot of duplicate code writing the type hints when defining an endpoint, for example:

```python
@router.post("/", response_model=FileRead)
async def upload_file(
    *,
    file: UploadFile,
    session: Session = Depends(get_session),
    user: OIDCUser = Depends(get_user),
    file_manager: AsnycBaseFileManager = Depends(get_file_manager),
) -> File:
    ...

@router.get("/content/{file_id}")
async def get_file_content(
    *,
    file_id: int,
    session: Session = Depends(get_session),
    file_manager: AsyncBaseFileManager = Depends(get_file_manager),
) -> StreamingResponse:
   ...
```

There is also room for human error if you write the wrong type hint, with no static type checking tools being able to prevent this.

![image](https://user-images.githubusercontent.com/44192816/164703476-0ffc0298-01b1-4be7-8f46-28d289178b2c.png)

It is possible to use Generics to properly type hint the Depends function:

```python
T = TypeVar("T")

def Depends(  # noqa: N802
    dependency: Optional[Callable[..., T]] = None, *, use_cache: bool = True
) -> T:
    return params.Depends(dependency=dependency, use_cache=use_cache)
```

This leads to automatically having the correct type hints:

![image](https://user-images.githubusercontent.com/44192816/164704235-43fe1ef5-4d2e-4745-beac-653a8cb40b0f.png)

Is there a reason Depends explicitly returns Any, instead of an unbounded TypeVar? If not I am happy to PR this and PR documentation if necessary.

### Wanted Solution

```
T = TypeVar("T")

def Depends(  # noqa: N802
    dependency: Optional[Callable[..., T]] = None, *, use_cache: bool = True
) -> T:
    return params.Depends(dependency=dependency, use_cache=use_cache)
```

### Wanted Code

```python
from fastapi import Depends, FastAPI

app = FastAPI()


def dependency() -> int:
    return 99


@app.get("/")
def read_root(my_int = Depends(dependency)):
    return {"an_int": my_int}
```

### Operating System

Windows

### FastAPI Version

0.75.2

### Python Version

3.9.12

**Comment by vrslev** [comment, marked as answer]
Hey there!

Technically it wouldn't be correct to do:
```python
def Depends(dependency: Optional[Callable[..., T]] = None, ...) -> T: ...
```
since the function actually returns `Depends` class. This shouldn't matter for end-user of the package, but can be confusing.

FastAPI supports async dependencies: in this case `T` would be type inferred as `Coroutine[Any, Any, T]`. The only solution would be to annotate an argument manually and add `type: ignore` comment.

Actually I found an issue that addresses similar problem: https://github.com/tiangolo/fastapi/issues/4750. It wouldn't be enough to add simple `TypeVar`—[you'd have to overloads](https://github.com/tiangolo/fastapi/issues/4750#issuecomment-1089185180).

**Comment by invokermain** [comment]
Ah I tried to find a previous issue but didn't find that one. I had a feeling that it wasn't that simple but seems like it would be a great benefit. Also interesting to learn about overloads, I've never had to use them yet. Will follow that issue thanks!

**Comment by tiangolo** [comment]
Thanks for the help here @vrslev ! 👏  🙇 

Thanks for reporting back and closing the issue @invokermain  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
